### PR TITLE
feat: Changes to storing and displaying transactions saved in the wallet (transaction versioning)

### PR DESCRIPTION
**Pull request by lszetg**
## Motivation and Context

I removed the migration steps based on our recent discussion. If migrating from `v1` to `v2` fails (e.g. `v2` version can't be fetched), the storage will still keep the `v1` version. However, on output, it will be temporarily converted to `WalletTransaction` to return a unified version.

I added gas to IOTA rebased and EVM in `v2`.

~~If this approach looks good, I’ll switch `v2` to `v1` and remove the extra version (i'll explain why it's unnecessary below).~~

Update: I think we can keep both v1 and v2 versions for now, as they provide a useful base for testing the migration logic. Maintaining both versions helps ensure that the migration path is working as expected and continues to be covered by tests.

## Summary

This proposal does not include migrating `wallet_transactions` to `wallet_transactions_versioned`.

I had trouble deciding where to do the migration so that it's invisible to the outside. One idea was to add a `migrate()` function to `UserEntity`:

```
impl UserEntity {
    pub fn migrate(&mut self) -> bool {
        if self.wallet_transactions.is_empty() {
            return false;
        }

        self.wallet_transactions_versioned = migrate_legacy_transactions_to_v1(self.wallet_transactions.clone());
        self.wallet_transactions = Vec::new();
        true
    }
}
```
and the most interesting place to call this function was inside `UserRepo::get()`:

```
fn get(&mut self, username: &str) -> Result<UserEntity> {
	debug!("Fetching entry in user DB");
	let mut u = self.inner.get(username)?;

	if !u.migrate() {
		self.update(&u)?;
	}

	Ok(u)
}
```

But this caused a few issues:
1. `get()` shouldn't mutate state
2. changing its signature required many updates in the SDK and bindings, forcing mut usage in functions with read locks, which didn’t look good 😅 

So current proposal: we don’t migrate (`legacy -> v1`) at all for now. Instead, re-fetch the data (something we already discussed on friday). Currently, there are no clients who would lose data, and re-fetching a few transactions is not a big issue.

## Checklist

Please ensure that your PR meets the following requirements:

### General

- [ ] Code follows project style and guidelines.
- [ ] No redundant or duplicate code.
- [ ] No added new dependencies without clear justification and approval.

### Documentation

- [ ] Added/updated relevant documentation (e.g., README, inline comments).
- [ ] Added a CHANGELOG entry for major changes.

### Testing

- [ ] Added relevant test cases, leveraging tools like `rstest` or similar for reusable tests.
- [ ] Tests pass locally.

### Build & CI/CD

- [ ] Confirmed no CI/CD pipeline issues.
- [ ] Updated build scripts where necessary.

## Comments to Reviewer

Please provide any comments or points of consideration for the reviewers, such as specific areas of focus, potential edge cases, or context that would aid the review process.

### Reviewer Responsibilities

- [ ] Code adheres to style and guidelines.
- [ ] All tests are appropriately covered and they pass.
- [ ] Consider performance and/or security impacts.
- [ ] Constructive feedback is provided.

### CODE REVIEW

File: sdk/src/core/wallet.rs [hunk 1]
```diff
@@ -525,87 +529,51 @@ impl Sdk {
...
+
+        // 3) confirm pending transactions
+        let confirmed_transaction_slice = confirm_pending_transactions(&wallet, &migrated_transaction_slice).await?;
+
+        // 4) and save the updated list back to the wallet.
+        merge_transactions(&mut wallet_transactions, confirmed_transaction_slice);
+        let _ = repo.set_wallet_transactions(&user.username, wallet_transactions.clone());
+
+        // Get updated transaction slice
+        let requested_transactions = get_transaction_slice(network_key.clone(), &wallet_transactions, start, limit);
```
Comment (lszetg) [inline sdk/src/core/wallet.rs:565]: oh, forgot to remove this line, we can use `confirmed_transaction_slice`

File: sdk/src/core/wallet_helpers/confirm_pending_transactions.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+/// This function takes a list of versioned wallet transactions and attempts to confirm their status.
+/// It returns the same list of transactions, with updated entries if the confirmation succeeds.
+///
+/// # Returns
+///
+/// Returns `Ok(Vec<VersionedWalletTransaction>)` containing the original transactions,
+/// with updated status where applicable.
+pub async fn confirm_pending_transactions(
+    wallet: &WalletBorrow<'_>,
+    transactions: &Vec<VersionedWalletTransaction>,
```
Comment (antbern) [inline sdk/src/core/wallet_helpers/confirm_pending_transactions.rs:16]: I suggest we take a mutable slice and mutate the contents in-place here instead of reallocating and cloning everything all the time:

```suggestion
    transactions: &mut [VersionedWalletTransaction],
```

we can then iterate mutably and update the elements one-by-one if needed 💯

File: sdk/src/core/wallet_helpers/hashes_to_transactions.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+/// Fetch new transaction hashes and convert them into versioned transactions.
+///
+/// This function retrieves transaction hashes starting from the given index (`start`) up to the specified `limit`,
+/// using the provided network key. If any of the fetched hashes are not present in the provided
+/// `wallet_transactions`, they are converted into `VersionedWalletTransaction` instances and returned.
+///
+/// # Returns
+///
+/// Returns `Ok(Vec<VersionedWalletTransaction>)` containing only the newly discovered transactio
+pub async fn fetch_new_hashes_and_turn_into_versioned_transactions(
```
Comment (antbern) [inline sdk/src/core/wallet_helpers/hashes_to_transactions.rs:15]: I do see the point in breaking some of this logic out in to separate functions, but at some level it feels like it is a bit difficult to follow the logic here when you need to mentally jump between functions (with very long names 😆 ). Just saying that we need to carefully manage this trade-off. Also there are no tests or anything here so the multiple functions does not serve that kind of purpose either.

File: sdk/src/core/wallet_helpers/merge.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    current_transactions: &mut Vec<VersionedWalletTransaction>,
+    incoming_transactions: Vec<VersionedWalletTransaction>,
+) {
+    for tx in incoming_transactions {
+        let (hash, network_key) = match &tx {
+            VersionedWalletTransaction::V1(v1) => (&v1.transaction_hash, &v1.network_key),
+            VersionedWalletTransaction::V2(v2) => (&v2.transaction_hash, &v2.network_key),
+        };
+
+        if let Some(existing) = current_transactions.iter_mut().find(|tx| match tx {
```
Comment (antbern) [inline sdk/src/core/wallet_helpers/merge.rs:23]: Why do we do this again when it is already done in the check for existing hashes? Seems a bit wasteful to me to do the check twice?

Also I see this function is called twice in the same function, once for the new hashes and once for `confirmed_transactions`. should'nt the confirming logic mutate the list in-place instead of requiring yet another (potentially) expensive merge call? let's not make it more complicated than it has to be!

File: sdk/src/core/wallet_helpers/merge.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            VersionedWalletTransaction::V1(v1) => v1.transaction_hash == *hash && v1.network_key == *network_key,
+            VersionedWalletTransaction::V2(v2) => v2.transaction_hash == *hash && v2.network_key == *network_key,
```
Comment (antbern) [inline sdk/src/core/wallet_helpers/merge.rs:25]: Perhaps we can add helper functions on the VersionedWalletTransaction also for network key and hash? It is used in a lot of places

File: sdk/src/core/wallet_helpers/migrate.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+/// For each outdated transaction, it attempts to fetch the latest version using `wallet.get_wallet_tx`.
+/// If successful, the transaction is updated; otherwise, the original (outdated) transaction is retained.
+///
+/// # Returns
+///
+/// Returns `Ok(Vec<VersionedWalletTransaction>)` containing the same set of transactions as provided,
+/// with entries migrated to the latest version where possible.
+pub async fn migrate_transactions(
+    wallet: &WalletBorrow<'_>,
+    transactions: &Vec<VersionedWalletTransaction>,
```
Comment (antbern) [inline sdk/src/core/wallet_helpers/migrate.rs:18]: We can take a mut slice here and do the update in-place instead here too 💯  (also see my comment on the merge function)
```suggestion
    transactions: &mut [VersionedWalletTransaction],
```

File: sdk/src/core/wallet.rs [hunk 7]
```diff
@@ -525,87 +529,51 @@ impl Sdk {
...
+
+        // 2) migrate transactions to the latest version if necessary,
+        let migrated_transaction_slice = migrate_transactions(&wallet, &transaction_slice).await?;
+
+        // 3) confirm pending transactions
+        let confirmed_transaction_slice = confirm_pending_transactions(&wallet, &migrated_transaction_slice).await?;
+
+        // 4) and save the updated list back to the wallet.
+        merge_transactions(&mut wallet_transactions, confirmed_transaction_slice);
+        let _ = repo.set_wallet_transactions(&user.username, wallet_transactions.clone());
```
Comment (antbern) [inline sdk/src/core/wallet.rs:562]: A bit wasteful to clone the _entire_ `Vec` here (it could be quite large) when we anyways only want to keep a few selected (with start and limit) around. If we do the clone of the ones we are interested in instead and _then_ call `set_wallet_transactions` no `clone` is needed 💯

**Review by antbern — APPROVED** [review]
Looking good! Let's go with this, nice work! 💯
